### ISSUE TITLE
Better examples

### ISSUE THREAD

**Issue by romainfrancois** [romainfrancois/parallelGroupBy#4]
Need reasonable examples. For example, I initially wrote the impls using this: 

```
babynames, c("sex", "name" )
```

but there are many many groups in each chunk, which makes the joining take a long time and therefore a parallel version not better, but not worse either than a serial version. 

Now considering babynames is organized like this: 

```
   year sex      name    n       prop
1  1880   F      Mary 7065 0.07238359
2  1880   F      Anna 2604 0.02667896
3  1880   F      Emma 2003 0.02052149
4  1880   F Elizabeth 1939 0.01986579
5  1880   F    Minnie 1746 0.01788843
6  1880   F  Margaret 1578 0.01616720
7  1880   F       Ida 1472 0.01508119
8  1880   F     Alice 1414 0.01448696
9  1880   F    Bertha 1320 0.01352390
10 1880   F     Sarah 1288 0.01319605
...
```

If I then group by `c("year", "sex")` I get fewer groups and so the joining is better: 

```
> microbenchmark(serial = make_index_serial(babynames,
+     c("year", "sex")), parallelReduce = make_index_parallel(babynames,
+     c("year", "sex ..." ... [TRUNCATED]
Unit: milliseconds
           expr       min        lq    median        uq       max neval
         serial  84.10732  84.10732  84.92999  85.75265  85.75265     2
 parallelReduce  31.05697  31.05697  33.14208  35.22719  35.22719     2
 tbb_concurrent 122.47935 122.47935 124.85138 127.22341 127.22341     2
          impl3  33.71695  33.71695  39.62045  45.52395  45.52395     2
```

which is somewhat reassuring. So I need better, realistic use cases. ping @hadley.

**Comment by hadley** [comment]
I wonder if for large datasets, we may need to do back to a pre-sort:

1) That would make joining simpler, because you could always split so that a group is only ever in one chunk
2) It will substantially improve locality for subsequent operations
3) We could use a simpler representation for the group indexes (as we tried out previously)

Maybe we could call sorting + group by sharding. It's going to be more expensive than just a group by, but it should make subsequent operations considerably faster.

It's also possible that babynames is just too small too to see the impact. We might need to get a server with more cores (e.g. > 10) and more memory (e.g. > 100) so that we can try it on larger sized data. I have three datasets that might be good candidates for testing, but we might need to work on fastread for a bit so its not such a pain to load them into memory.

**Comment by hadley** [comment]
Other idea: after sorting, you could do a parallel read through the data to figure out the number of groups, then preallocate index data structure, just two IntegerVectors (start & end, or start & length). I think that would avoid a copy, because when generating the index you could write in parallel without mutexes because every thread would be guaranteed to write to a different location.

**Comment by romainfrancois** [comment]
pre sort will be quite expensive on large data sets. one thing that is easy enough is testing whether it is already sorted in a first pass. when it is the case, there is just almost no joining to do. 

As for better representation of the indices, perhaps we could have two vectors: 
- a vector of size `n` where we have groups one after another
- a vector to hold the group sizes

As part of these benchmarks, sometimes it takes a "lot" of time to allocate the list and then its individual vectors, as I guess it leaves more chances for the GC to get involved. 

I'm doing some restructuring of this so that we can more easily test various implementations on various data ... on some machine.

**Comment by hadley** [comment]
I agree that a pre sort is going to be expensive, but if it isn't sorted, locality for subsequent summarise, mutate etc is going to be _horrible_.

If it does turn out to be important, I wonder if we could sort during load with fastread.

**Comment by romainfrancois** [comment]
Oh yeah, once we are in a sorted world, making group by parallel is much easier, but the total cost will definitely be dominated by the sort. 

Do people still save the result of `group_by` for subsequent uses ? This used to be "you do this one expensive thing once and then everything else is faster", but now I'm getting the feeling that people just write `%>%` stories as a whole. 

And yep, I do agree about the locality benefit. 

The only concern I'd have is about the toxicity of having both pre sorted and not pre sorted, e.g. how much changes in the dplyr code base would it take to handle both situations. 

I like the idea of presorting as the data gets in, I guess this is trivial if it comes from a database, for data that comes from files (e.g. fastread), I'm not sure this would be efficient to accumulate rows in a sorted fashion, with e.g a map. IIRC sorting at the end is more efficient than maintaining the data sorted, and we can probably take advantage of parallel sorting. tbb even has parallel_sort. 

That's a nice exercise altogether. For the time being, I just need to find a few realistic examples of group by. 

Would you say it is likely that the data is already organized by the grouping variables ?

**Comment by hadley** [comment]
I think things might play out a little differently for large datasets because things will be slower, so caching intermediate grouping might make more sense. I think we're going to have to play and see.

It's probably not terribly likely that the data is already sorted by grouping variables.
